sources/models/src/modeled_types/oci_defaults.rs

```
@@ -120,10 +157,155 @@ mod oci_defaults_rlimits {
+            ),
+            (MaxProcesses, "max-processes", "RLIMIT_NPROC"),
+            (
+                MaxRealtimePriority,
+                "max-realtime-priority",
+                "RLIMIT_RTPRIO",
+            ),
+            (MaxRealtimeTimeout, "max-realtime-timeout", "RLIMIT_RTTIME"),
+            (MaxResidentSet, "max-resident-set", "RLIMIT_RSS"),
+            (MaxStackSize, "max-stack-size", "RLIMIT_STACK"),
+        ];
+
+        for rlimit in rlimits {
+            check_rlimit_strings(rlimit);
+        }
+    }
+}
+
+/// OciDefaultsResourceLimitValue specifies what all values are
+/// allowed to be set for rlimit in default spec.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub enum OciDefaultsResourceLimitValue {
+    Value(u32),
+    Unlimited,
+}
+
+impl OciDefaultsResourceLimitValue {
+    pub fn to_u64(self) -> u64 {
+        match self {
+            OciDefaultsResourceLimitValue::Value(x) => x as u64,
+            OciDefaultsResourceLimitValue::Unlimited => u64::MAX,
+        }
+    }
+}
+
+impl Serialize for OciDefaultsResourceLimitValue {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match *self {
+            OciDefaultsResourceLimitValue::Value(x) => serializer.serialize_u32(x),
+            OciDefaultsResourceLimitValue::Unlimited => serializer.serialize_str("unlimited"),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for OciDefaultsResourceLimitValue {
+    fn deserialize<D>(
+        deserializer: D,
+    ) -> std::result::Result<OciDefaultsResourceLimitValue, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let json_value = Value::deserialize(deserializer)?;
+
+        return match &json_value {
+            Value::Number(number) => {
+                if number.is_u64() {
+                    let s = number.to_string().parse::<u32>().map_err(|e| {
+                        de::Error::custom(format!(
```

I believe we use `Snafu` for these errors in `modeled_types` so I would like to use that for these errors.

---

Serde expect errors while deserializing  in form of D::Error, hence we need to return the error in form of D::Error.

---

These custom errors only in Serde deserializers makes sense.